### PR TITLE
Backup restore

### PR DESCRIPTION
**Pull request by git-hyagi**
Adding backup logic (**no restore logic yet**).

It is pending the backup of s3|azure|sso secrets because we didn't implement them in pulp yet.

### CODE REVIEW

File: config/samples/simple.yaml [hunk 1]
```diff
@@ -33,7 +33,8 @@ spec:
   file_storage_storage_class: "standard"
   signing_secret: "signing-galaxy"
   signing_scripts_configmap: "signing-scripts"
-  image_web: "quay.io/pulp/galaxy-web"
+  image_web: "quay.io/pulp/pulp-web"
+  #image_web: "quay.io/pulp/galaxy-web"
```
Comment (fao89) [inline config/samples/simple.yaml:37]: why not delete it?

File: controllers/backup_pulp_dir.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	execCmd = []string{
+		"bash", "-c", "cp -fr /var/lib/pulp/. " + backupDir + "/pulp",
+	}
+	_, err = r.containerExec(pod, execCmd, pulpBackup.Name+"-backup-manager", pod.Namespace)
+	if err != nil {
+		log.Error(err, "Failed to backup pulp dir")
+		return err
+	}
+
+	log.Info("Pulp's directory backup finished!")
```
Comment (fao89) [inline controllers/backup_pulp_dir.go:34]:
```suggestion
	log.Info(pulpBackup.DeploymentType + "'s directory backup finished!")
```

File: api/v1alpha1/pulpbackup_types.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+// PulpBackupSpec defines the desired state of PulpBackup
+type PulpBackupSpec struct {
+	DeploymentType string `json:"deployment_type"`
+
+	// Name of the deployment to be backed up
+	// +kubebuilder:validation:Optional
+	DeploymentName string `json:"deployment_name"`
+
+	// +kubebuilder:default:="pulp"
+	PulpInstanceName string `json:"pulp_instance_name"`
```
Comment (fao89) [inline api/v1alpha1/pulpbackup_types.go:32]:
```suggestion
	InstanceName string `json:"instance_name"`
```

File: config/samples/repo-manager_v1alpha1_pulpbackup.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+apiVersion: repo-manager.pulpproject.org/v1alpha1
+kind: PulpBackup
+metadata:
+  name: pulpbackup-sample
+spec:
+  deployment_type: pulp
+  backup_storage_class: standard
+  pulp_instance_name: pulp
```
Comment (fao89) [inline config/samples/repo-manager_v1alpha1_pulpbackup.yaml:8]:
```suggestion
  instance_name: pulp
```

File: controllers/backup_cr.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+	// CR BACKUP
+	log.Info("Starting pulp CR backup process ...")
+	pulpSpec, _ := json.Marshal(pulp.Spec)
+	execCmd := []string{
+		"bash", "-c", "echo '" + string(pulpSpec) + "' > " + backupDir + "/cr_object",
+	}
+	_, err := r.containerExec(pod, execCmd, pulpBackup.Name+"-backup-manager", pod.Namespace)
+	if err != nil {
+		log.Error(err, "Failed to backup pulp CR")
```
Comment (fao89) [inline controllers/backup_cr.go:30]:
```suggestion
		log.Error(err, "Failed to backup" +  pulpBackup.DeploymentType + "CR")
```

File: controllers/pulp_status.go [hunk 6]
```diff
@@ -72,6 +72,7 @@ func (r *PulpReconciler) pulpStatus(ctx context.Context, pulp *repomanagerv1alph
 			Message:            "All tasks ran successfully",
 		})
 		r.Status().Update(ctx, pulp)
+		log.Info("Pulp operator finished execution ...")
```
Comment (fao89) [inline controllers/pulp_status.go:75]:
```suggestion
		log.Info(pulp.DeploymentType + " operator finished execution ...")
```

File: controllers/pulpbackup_controller.go [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+	}
+
+	err = r.backupPulpDir(ctx, pulpBackup, backupDir, pod)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	log.Info("Cleaning up backup resources ...")
+	r.cleanup(ctx, pulpBackup)
+	log.Info("Pulp CR Backup finished!")
```
Comment (fao89) [inline controllers/pulpbackup_controller.go:109]:
```suggestion
	log.Info(pulpBackup.DeploymentType + " CR Backup finished!")
```

File: controllers/pulpbackup_controller.go [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+	volumes := []corev1.Volume{{
+		Name: pulpBackup.Name + "-backup",
+		VolumeSource: corev1.VolumeSource{
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: pulpBackup.Name + "-backup-claim",
+			},
+		},
+	}}
+
+	if pulp.Spec.IsFileStorage {
```
Comment (fao89) [inline controllers/pulpbackup_controller.go:165]: it might change when we implement the object storage

**Review by fao89 — APPROVED** [review]
minor suggestions, but I'm approving it

File: controllers/pulpbackup_controller.go (hunk 8 above)
Comment (fao89) [reply controllers/pulpbackup_controller.go:165]:
```suggestion
	if len(pulp.Spec.ObjectStorageAzureSecret) == 0 && len(pulp.Spec.ObjectStorageS3Secret) == 0 {
```

File: config/samples/simple.yaml (hunk 1 above)
Comment (git-hyagi) [reply config/samples/simple.yaml:37]: My bad!
I was doing some tests and should not have even pushed this.

File: controllers/backup_cr.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+	log := ctrllog.FromContext(ctx)
+
+	// we are considering that pulp CR instance is running in the same namespace as pulpbackup and
+	// that there is only a single instance of pulp CR available
+	// we could also let users pass the name of pulp instance
+	pulp := &repomanagerv1alpha1.Pulp{}
+	r.Get(ctx, types.NamespacedName{Name: pulpBackup.Spec.InstanceName, Namespace: pulpBackup.Namespace}, pulp)
+
+	// CR BACKUP
+	log.Info("Starting pulp CR backup process ...")
```
Comment (fao89) [inline controllers/backup_cr.go:23]: s/pulp/DeploymentType

**Review by fao89 — APPROVED** [review]

**Pull request merged**